src/oni2-keyboard-layout/stubs/keyboard-layout-mac.c

```diff
@@ -0,0 +1,233 @@
+  );
+
+  if (status == noErr && deadKeyState != 0) {
+    status = UCKeyTranslate(
+      keyboardLayout,
+      (uint16_t)virtualKeyCode,
+      kUCKeyActionDown,
+      modifierKeyState,
+      LMGetKbdLast(),
+      kUCKeyTranslateNoDeadKeysBit,
+      &deadKeyState,
+      4 * sizeof(characters[0]),
+      &charCount,
+      characters
+    );
+  }
+
+  if (status == noErr && !iscntrl(characters[0])) {
+    return charCount;
+  } else {
+    return 0;
+  }
+}
+
+CAMLprim value oni2_KeyboardLayoutPopulateCurrentKeymap(value keymap, value Hashtbl_replace) {
+  CAMLparam2(keymap, Hashtbl_replace);
+  CAMLlocal2(keymapEntry, vSDLScancode);
+
+  // Allocate string pointers to be filled
+  const char *unmodified;
+  const char *withShift;
+  const char *withAltGraph;
+  const char *withAltGraphShift;
+
+  // Allocate UTF-16 Buffers to be populated
+  unichar uniUnmodified[4]; 
+  unichar uniWithShift[4]; 
+  unichar uniWithAltGraph[4]; 
+  unichar uniWithAltGraphShift[4]; 
+
+  TISInputSourceRef source = TISCopyCurrentKeyboardInputSource();
+  CFDataRef layoutData = (CFDataRef)TISGetInputSourceProperty(source, kTISPropertyUnicodeKeyLayoutData);
+
+  if (layoutData == NULL) {
+    CAMLreturn(Val_unit);
+  }
+
+  const UCKeyboardLayout *keyboardLayout = (UCKeyboardLayout *)CFDataGetBytePtr(layoutData);
+
+  size_t keyCodeMapSize = sizeof(keyCodeMap) / sizeof(keyCodeMap[0]);
+  for (size_t i = 0; i < keyCodeMapSize; i++) {
+    SDL_Scancode sdlScancode = keyCodeMap[i].sdlScancode;
+    int virtualKeyCode = keyCodeMap[i].virtualKeyCode;
+
+    if (sdlScancode && virtualKeyCode < 0xffff) {
+      int ccUnmodified = characterForNativeCode(
+        keyboardLayout,
+        virtualKeyCode,
+        0, 
+        uniUnmodified
+      );
+      int ccWithShift = characterForNativeCode(
+        keyboardLayout,
+        virtualKeyCode, 
+        (1 << shiftKeyBit), 
+        uniWithShift
+      );
+      int ccWithAltGraph = characterForNativeCode(
+        keyboardLayout,
+        virtualKeyCode,
+        (1 << optionKeyBit),
+        uniWithAltGraph
+      );
+      int ccWithAltGraphShift = characterForNativeCode(
+        keyboardLayout,
+        virtualKeyCode,
+        (1 << shiftKeyBit) | (1 << optionKeyBit),
+        uniWithAltGraphShift
+      );
+
+      // The NSStrings we allocate are simply mechanisms to convert the UTF-16 code points to UTF-8
+      // so they should be autoreleased.
+      @autoreleasepool {
+        unmodified = [[NSString stringWithCharacters:uniUnmodified length:ccUnmodified] UTF8String];
+        withShift = [[NSString stringWithCharacters:uniWithShift length:ccWithShift] UTF8String];
+        withAltGraph = [[NSString stringWithCharacters:uniWithAltGraph length:ccWithAltGraph] UTF8String];
+        withAltGraphShift = [[NSString stringWithCharacters:uniWithAltGraphShift length:ccWithAltGraphShift] UTF8String];
+      }
+      
+      keymapEntry = createKeymapEntry(unmodified, withShift, withAltGraph, withAltGraphShift);
```

Comment:
Hi @zbaylin , I just thought of something - is it possible for the `unmodified`, `withShift`, etc keys to be released _prior_ to them being used in `createKeymapEntry`? Just worried about a possible race condition here, where the strings could be released prior to being copied to OCaml objects.

Reply:
That's a good point. According to the `UTF8String` selector documentation:
> This C string is a pointer to a structure inside the string object, which may have a lifetime shorter than the string object and will certainly not have a longer lifetime.

So I think it makes sense to move that call to `createKeymapEntry` into the `autoreleasepool`, or manually create and release the `NSString`s. So sorry I didn't think of that before merging!